File: src/vanilla/DanModel/JavaFileContents.cs
using AutoRest.Core.Utilities;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AutoRest.Java.DanModel
{
    public class JavaFileContents
    {
        private const string singleIndent = "    ";

        private readonly StringBuilder contents = new StringBuilder();
        private readonly StringBuilder linePrefix = new StringBuilder();

        private int? wordWrapWidth;

        private bool previousLineEndingPending;

        public override string ToString()
        {
            return contents.ToString();
        }

        public string[] Lines
        {
            get { return ToString().Split("\n"); }
        }

        public void AddToPrefix(string toAdd)
        {
            linePrefix.Append(toAdd);
        }

        private void RemoveFromPrefix(string toRemove)
        {
            int toRemoveLength = toRemove.Length;
            if (linePrefix.Length <= toRemoveLength)
            {
                linePrefix.Clear();
            }
            else
            {
                linePrefix.Remove(linePrefix.Length - toRemoveLength, toRemoveLength);
            }
        }

        public void SetWordWrapWidth(int? wordWrapWidth)
        {
            this.wordWrapWidth = wordWrapWidth;
        }

        private void WithWordWrap(int wordWrapWidth, Action action)
        {
            SetWordWrapWidth(wordWrapWidth);
            action.Invoke();
            SetWordWrapWidth(null);
        }

        public void Indent(Action action)
        {
            IncreaseIndent();
            action.Invoke();
            DecreaseIndent();
        }

        public void IncreaseIndent()
            => AddToPrefix(singleIndent);

        public void DecreaseIndent()
            => RemoveFromPrefix(singleIndent);

        private IEnumerable<string> WordWrap(string line, bool addPrefix)
        {
            List<string> lines = new List<string>();

            if (wordWrapWidth == null)
            {
                lines.Add(line);
            }
            else
            {
                // Subtract an extra column from the word wrap width because columns generally are
                // 1 -based instead of 0-based.
                int wordWrapIndexMinusLinePrefixLength = wordWrapWidth.Value - (addPrefix ? linePrefix.Length : 0) - 1;
                IEnumerable<string> wrappedLines = line.WordWrap(wordWrapIndexMinusLinePrefixLength);
                foreach (string wrappedLine in wrappedLines.SkipLast(1))
                {
                    lines.Add(wrappedLine + "\n");
                }

                string lastWrappedLine = wrappedLines.Last();
                if (!string.IsNullOrEmpty(lastWrappedLine))
                {
                    lines.Add(lastWrappedLine);
                }
            }

            return lines;
        }

        private void Text(string text, bool addPrefix)
        {
            List<string> lines = new List<string>();

            if (string.IsNullOrEmpty(text))
            {
                lines.Add("");
            }
            else
            {
                int lineStartIndex = 0;
                int textLength = text.Length;
                while (lineStartIndex < textLength)
                {
                    int newLineCharacterIndex = text.IndexOf('\n', lineStartIndex);
                    if (newLineCharacterIndex == -1)
                    {
                        string line = text.Substring(lineStartIndex);
                        IEnumerable<string> wrappedLines = WordWrap(line, addPrefix);
                        lines.AddRange(wrappedLines);
                        lineStartIndex = textLength;
                    }
                    else
                    {
                        int nextLineStartIndex = newLineCharacterIndex + 1;
                        string line = text.Substring(lineStartIndex, nextLineStartIndex - lineStartIndex);
                        IEnumerable<string> wrappedLines = WordWrap(line, addPrefix);
                        lines.AddRange(wrappedLines);
                        lineStartIndex = nextLineStartIndex;
                    }
                }
            }

            string prefix = addPrefix ? linePrefix.ToString() : null;
            foreach (string line in lines)
            {
                if (addPrefix && !string.IsNullOrWhiteSpace(prefix) || (!string.IsNullOrEmpty(prefix) && !string.IsNullOrWhiteSpace(line)))
                {
                    contents.Append(prefix);
                }

                contents.Append(line);
            }
        }

        private void ClosePreviousLineEnding()
        {
            if (previousLineEndingPending)
            {
                previousLineEndingPending = false;

                Line();
            }
        }

        public void Text(string text)
        {
            ClosePreviousLineEnding();

            Text(text, addPrefix: true);
        }

        private void Line(string text, bool addPrefix)
        {
            Text($"{text}{Environment.NewLine}", addPrefix);
        }

        public JavaFileContents Line(string text, params object[] formattedArguments)
        {
            ClosePreviousLineEnding();

            if (formattedArguments != null && formattedArguments.Length > 0)
            {
                text = string.Format(text, formattedArguments);
            }

            Line(text, addPrefix: true);
            return this;
        }

        public JavaFileContents Line()
        {
            return Line("");
        }

        public void Package(string package)
        {
            Line($"package {package};");
        }

        public void Block(string text, Action<JavaBlock> bodyAction)
        {
            Line($"{text} {{");
            Indent(() =>
                {
                    bodyAction.Invoke(new JavaBlock(this));
                });
            Line($"}}");
        }

        public void BlockStatement(string text, Action<JavaBlock> bodyAction)
        {
            Line($"{text} {{");
            Indent(() =>
            {
                bodyAction.Invoke(new JavaBlock(this));
            });
            Line($"}};");
        }

        public void Import(params string[] imports)
        {
            Import((IEnumerable<string>)imports);
        }

        public void Import(IEnumerable<string> imports)
        {
            if (imports != null && imports.Any())
            {
                ISet<string> importSet = new SortedSet<string>(imports, new JavaImportComparer());
                foreach (string import in importSet)
                {
                    if (!string.IsNullOrEmpty(import))
                    {
                        Line($"import {import};");
                    }
                }
                Line();
            }
        }

        public void SingleLineSlashStarComment(string text)
        {
            Line($"/** {text} */");
        }

        public void SingleLineSlashSlashComment(string text)
        {
            Line($"// {text}");
        }

        public void SingleLineComment(string text)
        {
            SingleLineSlashStarComment(text);
        }

        public void MultipleLineSlashStarComment(Action<JavaMultipleLineComment> commentAction)
        {
            Line("/**");
            AddToPrefix(" * ");
            commentAction.Invoke(new JavaMultipleLineComment(this));
            RemoveFromPrefix(" * ");
            Line(" */");
        }

        public void MultipleLineSlashSlashComment(Action<JavaMultipleLineComment> commentAction)
        {
            AddToPrefix("// ");
            commentAction.Invoke(new JavaMultipleLineComment(this));
            RemoveFromPrefix("// ");
        }

        public void MultipleLineComment(Action<JavaMultipleLineComment> commentAction)
        {
            MultipleLineSlashStarComment(commentAction);
        }

        public void WordWrappedMultipleLineSlashStarComment(int wordWrapWidth, Action<JavaMultipleLineComment> commentAction)
        {
            MultipleLineComment((comment) =>
            {
                WithWordWrap(wordWrapWidth, () =>
                {
                    commentAction.Invoke(new JavaMultipleLineComment(this));
                });
            });
        }

        public void WordWrappedMultipleLineSlashSlashComment(int wordWrapWidth, Action<JavaMultipleLineComment> commentAction)
        {
            MultipleLineSlashSlashComment((comment) =>
            {
                WithWordWrap(wordWrapWidth, () =>
                {
                    commentAction.Invoke(new JavaMultipleLineComment(this));
                });
            });
        }

        public void WordWrappedMultipleLineComment(int wordWrapWidth, Action<JavaMultipleLineComment> commentAction)
        {
            WordWrappedMultipleLineSlashStarComment(wordWrapWidth, commentAction);
        }

        public void Return(string text)
        {
            Line($"return {text};");
        }

        public void ReturnBlock(string text, Action<JavaBlock> bodyAction)
        {
            Line($"return {text} {{");
            Indent(() =>
            {
                bodyAction.Invoke(new JavaBlock(this));
            });
            Line($"}};");
        }

        public void ReturnAnonymousClass(string anonymousClassDeclaration, Action<JavaClass> anonymousClassBlock)
        {
            Line($"return {anonymousClassDeclaration} {{");
            Indent(() =>
            {
                JavaClass javaClass = new JavaClass(this);
                anonymousClassBlock.Invoke(javaClass);
            });
            Line("};");
        }

        public void Annotation(params string[] annotations)
        {
            Annotation((IEnumerable<string>)annotations);
        }

        public void Annotation(IEnumerable<string> annotations)
        {
            if (annotations != null && annotations.Any())
            {
                foreach (string annotation in annotations)
                {
                    if (!string.IsNullOrEmpty(annotation))
                    {
                        Line($"@{annotation}");
                    }
                }
            }
        }

        public void PublicFinalClass(string classDeclaration, Action<JavaClass> classAction)
        {
            Block($"public final class {classDeclaration}", (blockAction) =>
            {
                if (classAction != null)
                {
                    JavaClass javaClass = new JavaClass(this);
                    classAction.Invoke(javaClass);
                }
            });
        }

        public void PublicClass(string classDeclaration, Action<JavaClass> classAction)
        {
            Block($"public class {classDeclaration}", (blockAction) =>
            {
                if (classAction != null)
                {
                    JavaClass javaClass = new JavaClass(this);
                    classAction.Invoke(javaClass);
                }
            });
        }

        public void PrivateStaticFinalClass(string classSignature, Action<JavaClass> classAction)
        {
            Block($"private static final class {classSignature}", (blockAction) =>
            {
                if (classAction != null)
                {
                    JavaClass javaClass = new JavaClass(this);
                    classAction.Invoke(javaClass);
                }
            });
        }

        public void PublicEnum(string enumName, Action<JavaBlock> enumAction)
        {
            Block($"public enum {enumName}", enumAction);
        }

        public void PublicInterface(string interfaceSignature, Action<JavaInterface> interfaceAction)
        {
            Line($"public interface {interfaceSignature} {{");
            Indent(() =>
            {
                interfaceAction.Invoke(new JavaInterface(this));
            });
            Line($"}}");
        }

        public void Interface(string interfaceSignature, Action<JavaInterface> interfaceAction)
        {
            Line($"interface {interfaceSignature} {{");
            Indent(() =>
            {
                interfaceAction.Invoke(new JavaInterface(this));
            });
            Line($"}}");
        }

        public void CommentParam(string parameterName, string parameterDescription)
        {
            Line($"@param {parameterName} {parameterDescription}");
        }

        public void CommentReturn(string returnValueDescription)
        {
            Line($"@return {returnValueDescription}");
        }

        public void CommentThrows(string exceptionTypeName, string description)
        {
            Line($"@throws {exceptionTypeName} {description}");
        }

        public void If(string condition, Action<JavaBlock> ifAction)
        {
            Line($"if ({condition}) {{");
            Indent(() =>
            {
                ifAction.Invoke(new JavaBlock(this));
            });
            Text($"}}");

            previousLineEndingPending = true;
        }

        public void ElseIf(string condition, Action<JavaBlock> elseIfAction)
        {
            previousLineEndingPending = false;

            Line($" else if ({condition}) {{", addPrefix: false);
            Indent(() =>
            {
                elseIfAction.Invoke(new JavaBlock(this));
            });
            Text($"}}");

            previousLineEndingPending = true;
        }

        public void Else(Action<JavaBlock> elseAction)
        {
            previousLineEndingPending = false;

            Line($" else {{", addPrefix: false);
            Indent(() =>
            {
                elseAction.Invoke(new JavaBlock(this));
            });
            Line($"}}");
        }
    }
}

File: javagen/src/main/java/com/azure/autorest/model/javamodel/JavaFileContents.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.autorest.model.javamodel;

import com.azure.core.util.CoreUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;
import java.util.function.Consumer;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class JavaFileContents {
    private static final String SINGLE_INDENT = "    ";
    private static final Pattern QUOTED_NEW_LINE = Pattern.compile(Pattern.quote("\n"));

    private final StringBuilder contents;
    private final StringBuilder linePrefix;

    private CurrentLineType currentLineType = CurrentLineType.values()[0];

    public JavaFileContents() {
        this(null);
    }

    public JavaFileContents(String fileContents) {
        contents = new StringBuilder();
        linePrefix = new StringBuilder();

        if (fileContents != null && !fileContents.isEmpty()) {
            contents.append(fileContents);
        }
    }

    private static String toString(List<JavaModifier> modifiers) {
        return modifiers == null ? "" : modifiers.stream().map(JavaModifier::toString).collect(Collectors.joining(" "));
    }

    @Override
    public String toString() {
        return contents.toString();
    }

    public final String[] getLines() {
        return QUOTED_NEW_LINE.split(toString(), -1);
    }

    public final void addToPrefix(String toAdd) {
        linePrefix.append(toAdd);
    }

    private void removeFromPrefix(String toRemove) {
        int toRemoveLength = toRemove.length();
        if (linePrefix.length() <= toRemoveLength) {
            linePrefix.setLength(0);
        } else {
            linePrefix.delete(linePrefix.length() - toRemoveLength, linePrefix.length() - toRemoveLength + toRemoveLength);
        }
    }

    public final void indent(Runnable action) {
        increaseIndent();
        action.run();
        decreaseIndent();
    }

    public final void increaseIndent() {
        addToPrefix(SINGLE_INDENT);
    }

    public final void decreaseIndent() {
        removeFromPrefix(SINGLE_INDENT);
    }

    private void text(String text, boolean addPrefix) {
        ArrayList<String> lines = new ArrayList<>();

        if (text == null || text.isEmpty()) {
            lines.add("");
        } else {
            int lineStartIndex = 0;
            int textLength = text.length();
            while (lineStartIndex < textLength) {
                int newLineCharacterIndex = text.indexOf('\n', lineStartIndex);
                if (newLineCharacterIndex == -1) {
                    String line = text.substring(lineStartIndex);
                    lines.add(line);
                    lineStartIndex = textLength;
                } else {
                    int nextLineStartIndex = newLineCharacterIndex + 1;
                    String line = text.substring(lineStartIndex, nextLineStartIndex);
                    lines.add(line);
                    lineStartIndex = nextLineStartIndex;
                }
            }
        }

        String prefix = addPrefix ? linePrefix.toString() : null;
        for (String line : lines) {
            if (addPrefix && prefix != null && !prefix.trim().isEmpty() || (prefix != null && !prefix.isEmpty() && line != null && !line.trim().isEmpty())) {
                contents.append(prefix);
            }

            contents.append(line);
        }
    }

    public final void text(String text) {
        if (currentLineType == CurrentLineType.Empty) {
            text(text, true);
        } else if (currentLineType == CurrentLineType.Text) {
            text(text, false);
        } else if (currentLineType == CurrentLineType.AfterIf) {
            line("", false);
            text(text, true);
        }
        currentLineType = CurrentLineType.Text;
    }

    private void line(String text, boolean addPrefix) {
        text(text + "\n", addPrefix);
        currentLineType = CurrentLineType.Empty;
    }

    public void line(String text) {
        if (currentLineType == CurrentLineType.Empty) {
            line(text, true);
        } else if (currentLineType == CurrentLineType.Text) {
            line(text, false);
        } else if (currentLineType == CurrentLineType.AfterIf) {
            line("", false);
            line(text, true);
        }
        currentLineType = CurrentLineType.Empty;
    }

    public void line(String text, Object... formattedArguments) {
        if (formattedArguments != null && formattedArguments.length > 0) {
            text = String.format(text, formattedArguments);
        }

        line(text);
    }

    public void line() {
        line("");
    }

    public void declarePackage(String pkg) {
        line("package " + pkg + ";");
    }

    public void block(String text, Consumer<JavaBlock> bodyAction) {
        line(text + " {");
        indent(() ->
                bodyAction.accept(new JavaBlock(this)));
        line("}");
    }

    public void declareImport(String... imports) {
        declareImport(Arrays.asList(imports));
    }

    public void declareImport(List<String> imports) {
        if (imports != null && !imports.isEmpty()) {
            Set<String> importSet = new TreeSet<>(new JavaImportComparer());
            importSet.addAll(imports);
            for (String toImport : importSet) {
                if (toImport != null && !toImport.isEmpty()) {
                    line("import " + toImport + ";");
                }
            }
            line();
        }
    }

    public void lineComment(String text) {
        lineComment(comment -> comment.line(text));
    }

    public void lineComment(Consumer<JavaLineComment> commentAction) {
        addToPrefix("// ");
        commentAction.accept(new JavaLineComment(this));
        removeFromPrefix("// ");
    }

    public void blockComment(String text) {
        blockComment(comment -> comment.line(text));
    }

    public void blockComment(Consumer<JavaLineComment> commentAction) {
        line("/*");
        addToPrefix(" * ");
        commentAction.accept(new JavaLineComment(this));
        removeFromPrefix(" * ");
        line(" */");
    }

    public void javadocComment(String text) {
        javadocComment(comment -> comment.description(text));
    }

    public void javadocComment(Consumer<JavaJavadocComment> commentAction) {
        line("/**");
        addToPrefix(" * ");
        commentAction.accept(new JavaJavadocComment(this));
        removeFromPrefix(" * ");
        line(" */");
    }

    public void methodReturn(String text) {
        line("return " + text + ";");
    }

    public void returnAnonymousClass(String anonymousClassDeclaration, Consumer<JavaClass> anonymousClassBlock) {
        line("return " + anonymousClassDeclaration + " {");
        indent(() -> {
            JavaClass javaClass = new JavaClass(this);
            anonymousClassBlock.accept(javaClass);
        });
        line("};");
    }

    public void anonymousClass(String anonymousClassDeclaration, String instanceName, Consumer<JavaClass> anonymousClassBlock) {
        line(anonymousClassDeclaration + " " + instanceName + " = new " + anonymousClassDeclaration + "() {");
        indent(() -> {
            JavaClass javaClass = new JavaClass(this);
            anonymousClassBlock.accept(javaClass);
        });
        line("};");
    }

    public void annotation(String... annotations) {
        annotation(Arrays.asList(annotations));
    }

    public void annotation(List<String> annotations) {
        if (annotations != null && !annotations.isEmpty()) {
            for (String annotation : annotations) {
                if (annotation != null && !annotation.isEmpty()) {
                    line("@" + annotation);
                }
            }
        }
    }

    public void classBlock(JavaVisibility visibility, List<JavaModifier> modifiers, String classDeclaration,
        Consumer<JavaClass> classAction) {
        String text = CoreUtils.isNullOrEmpty(modifiers)
            ? visibility + " class " + classDeclaration
            : visibility + " " + toString(modifiers) + " class " + classDeclaration;
        block(text, blockAction -> {
            if (classAction != null) {
                JavaClass javaClass = new JavaClass(this);
                classAction.accept(javaClass);
            }
        });
    }

    public void method(JavaVisibility visibility, List<JavaModifier> modifiers, String methodSignature,
        Consumer<JavaBlock> method) {
        String text = CoreUtils.isNullOrEmpty(modifiers)
            ? visibility + " " + methodSignature
            : visibility + " " + toString(modifiers) + " " + methodSignature;

        block(text, method);
    }

    public void constructor(JavaVisibility visibility, String constructorSignature, Consumer<JavaBlock> constructor) {
        block(visibility + " " + constructorSignature, constructor);
    }

    public void enumBlock(JavaVisibility visibility, String enumName, Consumer<JavaEnum> enumAction) {
        block(visibility + " enum " + enumName, block -> {
            if (enumAction != null) {
                JavaEnum javaEnum = new JavaEnum(this);
                enumAction.accept(javaEnum);
                javaEnum.addExpectedNewLineAfterLastValue();
            }
        });
    }

    public void interfaceBlock(JavaVisibility visibility, String interfaceSignature, Consumer<JavaInterface> interfaceAction) {
        line(visibility + " interface " + interfaceSignature + " {");
        indent(() -> interfaceAction.accept(new JavaInterface(this)));
        line("}");
    }

    public void ifBlock(String condition, Consumer<JavaBlock> ifAction) {
        line("if (" + condition + ") {");
        indent(() -> ifAction.accept(new JavaBlock(this)));
        text("}");
        currentLineType = CurrentLineType.AfterIf;
    }

    public void elseIfBlock(String condition, Consumer<JavaBlock> ifAction) {
        line(" else if (" + condition + ") {", false);
        indent(() -> ifAction.accept(new JavaBlock(this)));
        text("}");
        currentLineType = CurrentLineType.AfterIf;
    }

    public void elseBlock(Consumer<JavaBlock> elseAction) {
        line(" else {", false);
        indent(() -> elseAction.accept(new JavaBlock(this)));
        line("}");
    }

    public void tryBlock(Consumer<JavaBlock> tryAction) {
        line("try {");
        indent(() -> tryAction.accept(new JavaBlock(this)));
        text("}");
        currentLineType = CurrentLineType.AfterIf;
    }

    public void tryBlock(String resource, Consumer<JavaBlock> tryAction) {
        line("try (" + resource + ") {");
        indent(() -> tryAction.accept(new JavaBlock(this)));
        text("}");
        currentLineType = CurrentLineType.AfterIf;
    }

    public void catchBlock(String exception, Consumer<JavaBlock> catchAction) {
        line(" catch (" + exception + ") {", false);
        indent(() -> catchAction.accept(new JavaBlock(this)));
        line("}");
        currentLineType = CurrentLineType.AfterIf;
    }

    public void finallyBlock(Consumer<JavaBlock> finallyAction) {
        line(" finally {", false);
        indent(() -> finallyAction.accept(new JavaBlock(this)));
        line("}");
    }

    public void lambda(String parameterType, String parameterName, Consumer<JavaLambda> body) {
        text("(" + parameterType + " " + parameterName + ") -> ");
        try (JavaLambda lambda = new JavaLambda(this)) {
            body.accept(lambda);
        }
    }

    public void lambda(String parameterType, String parameterName, String returnExpression) {
        lambda(parameterType, parameterName, lambda -> lambda.lambdaReturn(returnExpression));
    }

    private enum CurrentLineType {
        Empty,
        AfterIf,
        Text
    }
}
